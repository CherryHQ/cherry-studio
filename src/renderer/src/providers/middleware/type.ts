import { MCPTool, MCPToolResponse, Metrics, Usage, WebSearchResponse } from '@renderer/types'
import { Chunk, ErrorChunk } from '@renderer/types/chunk'

import { BaseApiClient, SdkParams, SdkRawChunk, SdkRawOutput, SdkToolCall } from '../AiProvider/clients'
import { CompletionsParams, GenericChunk, OnFilterMessagesFunction } from './schemas'

/**
 * Symbol to uniquely identify middleware context objects.
 */
export const MIDDLEWARE_CONTEXT_SYMBOL = Symbol.for('AiProviderMiddlewareContext')

/**
 * Defines the structure for the onChunk callback function.
 */
export type OnChunkFunction = (chunk: Chunk | ErrorChunk) => void

/**
 * Base context that carries information about the current method call.
 */
export interface BaseContext {
  [MIDDLEWARE_CONTEXT_SYMBOL]: true
  methodName: string
  apiClientInstance: BaseApiClient
  originalParams: Readonly<CompletionsParams>

  readonly onChunkCallback: (chunk: Chunk) => void
  readonly onFilterMessagesCallback?: OnFilterMessagesFunction // 可选
}

/**
 * Processing state shared between middlewares.
 */
export interface ProcessingState {
  sdkPayload?: Readonly<SdkParams>
  capabilities?: {
    isStreaming: boolean
    isEnabledToolCalling: boolean
    isEnabledWebSearch: boolean
    isEnabledReasoning: boolean
    mcpTools: MCPTool[]
  }
  observer?: {
    usage?: Usage
    metrics?: Metrics
  }
  apiCall?: {
    requestTimestamp?: number
    rawSdkOutput?: SdkRawOutput
    rawSdkStream?: ReadableStream<SdkRawChunk> // Output from StreamAdapterMiddleware, consumed by SdkChunkToGenericChunkMiddleware
    genericChunkStream?: ReadableStream<GenericChunk> // Output from SdkChunkToGenericChunkMiddleware, consumed by Generic processors (Text, Think, Image, etc.)
  }
  toolProcessingState?: {
    pendingToolCalls: Array<SdkToolCall>
    executingToolCalls: Array<{
      sdkToolCall: SdkToolCall
      mcpToolResponse: MCPToolResponse
    }>
    isRecursiveCall?: boolean
    recursionDepth?: number
  }
  webSearchState?: {
    results?: WebSearchResponse[]
  }
  flowControl?: {
    abortController?: AbortController
    abortSignal?: AbortSignal
    cleanup?: () => void
  }
  customState?: Record<string, any>
}

/**
 * Extended context for completions method.
 */
export interface CompletionsContext extends BaseContext {
  readonly methodName: 'completions' // 强制方法名为 'completions'

  // --- Mutable internal state for the duration of the middleware chain ---
  _internal: ProcessingState // 包含所有可变的处理状态
}

/**
 * Next function type in Koa style.
 */
export type Next = () => Promise<void>

/**
 * Base middleware type.
 */
export type Middleware<TContext extends BaseContext> = (ctx: TContext, next: Next) => Promise<void>

/**
 * Completions middleware type.
 */
export type CompletionsMiddleware = Middleware<CompletionsContext>

/**
 * Generic method middleware type.
 */
export type MethodMiddleware = Middleware<BaseContext>

/**
 * Base configuration for any middleware.
 */
export interface BaseMiddlewareConfig {
  id?: string
  name?: string
}

/**
 * Middleware configuration structure.
 */
export interface MiddlewareConfig extends BaseMiddlewareConfig {
  completions?: CompletionsMiddleware[]
  methods?: Record<string, MethodMiddleware[]>
}

// Re-export for convenience
export type { Chunk as OnChunkArg } from '@renderer/types/chunk'
