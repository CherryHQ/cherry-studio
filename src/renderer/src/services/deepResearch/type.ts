import { z } from 'zod'

// Schema
// =====================================================
export const ClarifyWithUserSchema = z
  .object({
    need_clarification: z.boolean().describe('Whether the query needs clarification from the user'),
    question: z.string().describe('question to ask the user to clarify the report scope'),
    verification: z.string().describe('verification message that we will start research')
  })
  .transform((r) => ({
    needClarification: r.need_clarification,
    question: r.question,
    verification: r.verification
  }))
export type ClarifyWithUserResponse = z.infer<typeof ClarifyWithUserSchema>

export const ResearchTopicsSchema = z.object({
  reflection: z.string().describe('thorough reflection on the research brief and current findings'),
  tasks: z.array(z.string()).describe('list of research tasks to explore'),
  fulfilled: z.boolean().describe('whether the research brief is fulfilled by the current findings')
})
export type ResearchTopicsResponse = z.infer<typeof ResearchTopicsSchema>
// =====================================================
