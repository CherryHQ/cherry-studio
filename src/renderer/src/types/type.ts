export type Maybe<T> = T | undefined

/**
 * 获取对象的所有键名，并保持类型安全
 * @param obj - 要获取键名的对象
 * @returns 对象的所有键名数组，类型为对象键名的联合类型
 * @example
 * ```ts
 * const obj = { foo: 1, bar: 'hello' };
 * const keys = objectKeys(obj); // ['foo', 'bar']
 * ```
 */
export function objectKeys<T extends object>(obj: T): (keyof T)[] {
  return Object.keys(obj) as (keyof T)[]
}

/**
 * 将对象转换为键值对数组，保持类型安全
 * @template T - 对象类型
 * @param obj - 要转换的对象
 * @returns 键值对数组，每个元素是一个包含键和值的元组
 * @example
 * const obj = { name: 'John', age: 30 };
 * const entries = objectEntries(obj); // [['name', 'John'], ['age', 30]]
 */
export function objectEntries<T extends object>(obj: T): [keyof T, T[keyof T]][] {
  return Object.entries(obj) as [keyof T, T[keyof T]][]
}

/**
 * 将对象转换为键值对数组，提供更严格的类型检查
 * @template T - 对象类型，键必须是string、number或symbol，值可以是任意类型
 * @param obj - 要转换的对象
 * @returns 键值对数组，每个元素是一个包含键和值的元组，类型完全对应原对象的键值类型
 * @example
 * const obj = { name: 'John', age: 30 };
 * const entries = objectEntriesStrict(obj); // [['name', string], ['age', number]]
 */
export function objectEntriesStrict<T extends Record<string | number | symbol, unknown>>(
  obj: T
): { [K in keyof T]: [K, T[K]] }[keyof T][] {
  return Object.entries(obj) as { [K in keyof T]: [K, T[K]] }[keyof T][]
}

/**
 * 获取对象所有值的类型安全版本
 * @template T - 对象类型
 * @param obj - 要获取值的对象
 * @returns 对象值组成的数组
 * @example
 * const obj = { a: 1, b: 2 } as const;
 * const values = objectValues(obj); // (1 | 2)[]
 */
export function objectValues<T extends Record<string, unknown>>(obj: T): T[keyof T][] {
  return Object.values(obj) as T[keyof T][]
}

/**
 * 表示一个对象类型，该对象至少包含类型T中指定的所有键，这些键的值类型为U
 * 同时也允许包含其他任意string类型的键，这些键的值类型也必须是U
 * @template T - 必需包含的键的字面量字符串联合类型
 * @template U - 所有键对应值的类型
 * @example
 * type Example = AtLeast<'a' | 'b', number>;
 * // 结果类型允许:
 * const obj1: Example = { a: 1, b: 2 };           // 只包含必需的键
 * const obj2: Example = { a: 1, b: 2, c: 3 };     // 包含额外的键
 */
export type AtLeast<T extends string, U> = {
  [K in T]: U
} & {
  [key: string]: U
}

/**
 * 从对象中移除指定的属性键，返回新对象
 * @template T - 源对象类型
 * @template K - 要移除的属性键类型，必须是T的键
 * @param obj - 源对象
 * @param keys - 要移除的属性键列表
 * @returns 移除指定属性后的新对象
 * @example
 * ```ts
 * const obj = { a: 1, b: 2, c: 3 };
 * const result = strip(obj, ['a', 'b']);
 * // result = { c: 3 }
 * ```
 */
export function strip<T, K extends keyof T>(obj: T, keys: K[]): Omit<T, K> {
  const result = { ...obj }
  for (const key of keys) {
    delete (result as any)[key] // 类型上 Omit 已保证安全
  }
  return result
}

/**
 * Makes specified properties required while keeping others as is
 * @template T - The object type to modify
 * @template K - Keys of T that should be required
 * @example
 * type User = {
 *   name?: string;
 *   age?: number;
 * }
 *
 * type UserWithName = RequireSome<User, 'name'>
 * // Result: { name: string; age?: number; }
 */
export type RequireSome<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>
