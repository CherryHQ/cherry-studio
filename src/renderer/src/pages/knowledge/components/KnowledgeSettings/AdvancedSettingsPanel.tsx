import { InfoTooltip, Input } from '@cherrystudio/ui'
import type { KnowledgeBase } from '@renderer/types'
import type { ChangeEvent } from 'react'
import { useTranslation } from 'react-i18next'

interface AdvancedSettingsPanelProps {
  newBase: KnowledgeBase
  handlers: {
    handleChunkSizeChange: (value: number | null) => void
    handleChunkOverlapChange: (value: number | null) => void
    handleThresholdChange: (value: number | null) => void
  }
}

const AdvancedSettingsPanel: React.FC<AdvancedSettingsPanelProps> = ({ newBase, handlers }) => {
  const { t } = useTranslation()
  const { handleChunkSizeChange, handleChunkOverlapChange, handleThresholdChange } = handlers

  const handleNumericInputChange = (
    event: ChangeEvent<HTMLInputElement>,
    onChange: (value: number | null) => void,
    options: { min?: number; max?: number; allowFloat?: boolean } = {}
  ) => {
    const { value } = event.target

    if (value === '') {
      onChange(null)
      return
    }

    const parsed = options.allowFloat ? Number.parseFloat(value) : Number.parseInt(value, 10)
    if (!Number.isFinite(parsed)) {
      return
    }

    let sanitized = parsed

    if (!options.allowFloat) {
      sanitized = Math.round(sanitized)
    }

    if (options.min !== undefined) {
      sanitized = Math.max(options.min, sanitized)
    }

    if (options.max !== undefined) {
      sanitized = Math.min(options.max, sanitized)
    }

    onChange(sanitized)
  }

  return (
    <div className="px-4">
      <div className="mb-6">
        <div className="mb-2 flex items-center gap-2 text-sm">
          {t('knowledge.chunk_size')}
          <InfoTooltip content={t('knowledge.chunk_size_tooltip')} placement="right" />
        </div>
        <Input
          className="w-full"
          inputMode="numeric"
          min={100}
          step={1}
          value={newBase.chunkSize ?? ''}
          placeholder={t('knowledge.chunk_size_placeholder')}
          aria-label={t('knowledge.chunk_size')}
          onChange={(event) => handleNumericInputChange(event, handleChunkSizeChange, { min: 100 })}
        />
      </div>

      <div className="mb-6">
        <div className="mb-2 flex items-center gap-2 text-sm">
          {t('knowledge.chunk_overlap')}
          <InfoTooltip content={t('knowledge.chunk_overlap_tooltip')} placement="right" />
        </div>
        <Input
          className="w-full"
          inputMode="numeric"
          min={0}
          step={1}
          value={newBase.chunkOverlap ?? ''}
          placeholder={t('knowledge.chunk_overlap_placeholder')}
          aria-label={t('knowledge.chunk_overlap')}
          onChange={(event) => handleNumericInputChange(event, handleChunkOverlapChange, { min: 0 })}
        />
      </div>

      <div className="mb-6">
        <div className="mb-2 flex items-center gap-2 text-sm">
          {t('knowledge.threshold')}
          <InfoTooltip content={t('knowledge.threshold_tooltip')} placement="right" />
        </div>
        <Input
          className="w-full"
          inputMode="decimal"
          step={0.1}
          min={0}
          max={1}
          value={newBase.threshold ?? ''}
          placeholder={t('knowledge.threshold_placeholder')}
          aria-label={t('knowledge.threshold')}
          onChange={(event) =>
            handleNumericInputChange(event, handleThresholdChange, { min: 0, max: 1, allowFloat: true })
          }
        />
      </div>

      {/* <Alert className="border border-default-200 bg-default-50 text-sm" style={{ color: 'var(--color-warning)' }}>
        <AlertDescription className="text-current">
          {t('knowledge.chunk_size_change_warning')}
        </AlertDescription>
      </Alert> */}
    </div>
  )
}

export default AdvancedSettingsPanel
