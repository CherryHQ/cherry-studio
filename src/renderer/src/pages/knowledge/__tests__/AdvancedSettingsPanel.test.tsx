import type { KnowledgeBase, Model } from "@renderer/types";
import { fireEvent, render, screen } from "@testing-library/react";
import { beforeEach, describe, expect, it, vi } from "vitest";

import AdvancedSettingsPanel from "../components/KnowledgeSettings/AdvancedSettingsPanel";

const mocks = vi.hoisted(() => {
  return {
    i18n: {
      t: (k: string) => {
        const translations: Record<string, string> = {
          "knowledge.chunk_size": "分块大小",
          "knowledge.chunk_overlap": "分块重叠",
          "knowledge.threshold": "检索相似度阈值",
          "knowledge.chunk_size_change_warning": "避免修改这个高级设置。",
        };
        return translations[k] || k;
      },
    },
    handlers: {
      handleChunkSizeChange: vi.fn(),
      handleChunkOverlapChange: vi.fn(),
      handleThresholdChange: vi.fn(),
    },
  };
});

vi.mock("@cherrystudio/ui", () => ({
  InfoTooltip: ({ title }: { title: string }) => (
    <div>{mocks.i18n.t(title)}</div>
  ),
  Input: ({
    value,
    onChange,
    placeholder,
    "aria-label": ariaLabel,
    min,
    max,
    step,
    inputMode,
    className,
  }: any) => (
    <input
      type={
        inputMode === "numeric" || inputMode === "decimal" ? "number" : "text"
      }
      data-testid="input"
      className={className}
      aria-label={ariaLabel}
      placeholder={placeholder}
      value={value ?? ""}
      min={min}
      max={max}
      step={step}
      onChange={onChange}
    />
  ),
}));

vi.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: mocks.i18n.t,
  }),
}));

vi.mock("lucide-react", () => ({
  TriangleAlert: () => <span>warning</span>,
  CheckIcon: () => <span>check</span>,
  CircleXIcon: () => <span>error</span>,
  AlertTriangleIcon: () => <span>alert</span>,
}));

/**
 * 创建测试用的 KnowledgeBase 对象
 * @param overrides 可选的属性覆盖
 * @returns KnowledgeBase 对象
 */
function createKnowledgeBase(
  overrides: Partial<KnowledgeBase> = {},
): KnowledgeBase {
  return {
    id: "1",
    name: "Test KB",
    model: {
      id: "test-model",
      provider: "test-provider",
      name: "Test Model",
      group: "test",
    } as Model,
    items: [],
    created_at: Date.now(),
    updated_at: Date.now(),
    version: 1,
    chunkSize: 500,
    chunkOverlap: 200,
    threshold: 0.5,
    ...overrides,
  };
}

describe("AdvancedSettingsPanel", () => {
  const mockBase = createKnowledgeBase();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("basic rendering", () => {
    it("should match snapshot", () => {
      const { container } = render(
        <AdvancedSettingsPanel newBase={mockBase} handlers={mocks.handlers} />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("handlers", () => {
    it("should call handlers when values are changed", () => {
      render(
        <AdvancedSettingsPanel newBase={mockBase} handlers={mocks.handlers} />,
      );

      const inputs = screen.getAllByTestId("input");

      // Test chunk size change (first input)
      fireEvent.change(inputs[0], { target: { value: "600" } });
      expect(mocks.handlers.handleChunkSizeChange).toHaveBeenCalled();

      // Test chunk overlap change (second input)
      fireEvent.change(inputs[1], { target: { value: "300" } });
      expect(mocks.handlers.handleChunkOverlapChange).toHaveBeenCalled();

      // Test threshold change (third input)
      fireEvent.change(inputs[2], { target: { value: "0.6" } });
      expect(mocks.handlers.handleThresholdChange).toHaveBeenCalled();
    });
  });
});
