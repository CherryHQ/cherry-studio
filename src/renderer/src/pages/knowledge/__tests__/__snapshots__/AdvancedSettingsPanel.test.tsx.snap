// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvancedSettingsPanel > basic rendering > should match snapshot 1`] = `
.c0 {
  padding: 0 16px;
}

.c1 {
  margin-bottom: 24px;
}

.c1 .settings-label {
  font-size: 14px;
  margin-bottom: 8px;
  display: flex;
  align-items: center;
  gap: 8px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      分块大小
      <div />
    </div>
    <input
      aria-label="分块大小"
      data-testid="number-input"
      min="100"
      placeholder="knowledge.chunk_size_placeholder"
      type="number"
      value="500"
    />
  </div>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      分块重叠
      <div />
    </div>
    <input
      aria-label="分块重叠"
      data-testid="number-input"
      min="0"
      placeholder="knowledge.chunk_overlap_placeholder"
      type="number"
      value="200"
    />
  </div>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      检索相似度阈值
      <div />
    </div>
    <input
      aria-label="检索相似度阈值"
      data-testid="number-input"
      max="1"
      min="0"
      placeholder="knowledge.threshold_placeholder"
      step="0.1"
      type="number"
      value="0.5"
    />
  </div>
  <div
    data-color="warning"
    role="alert"
  >
    <span>
      warning
    </span>
    <strong>
      避免修改这个高级设置。
    </strong>
  </div>
</div>
`;
